src/pybind/mgr/dashboard/frontend/e2e/cluster/hosts.po.ts

```diff
@@ -1,5 +1,49 @@
+import { by, element } from 'protractor';
 import { PageHelper } from '../page-helper.po';
 
 export class HostsPageHelper extends PageHelper {
   pages = { index: '/#/hosts' };
+
+  // function that checks all services links work for first
+  // host in table (if a host is present)
+  check_services_links() {
+    this.navigateTo();
+    let links_tested = 0;
+    this.getTableCount()
+      .getText()
+      .then((hostcount) => {
+        // Only check for services links if there's at least 1 host present
+        if (hostcount.includes('0 total')) {
+          return;
+        }
+        // grab services column for first host
+        const services = element.all(by.css('.datatable-body-cell')).get(1);
+        // check is any services links are present
+        services.getText().then((txt) => {
+          // check that text (links) is present in services box
+          if (txt.length === 0) {
```

Comment:
Please use expect to validate this.
```suggestion
          expect(txt.length).toBeGreaterThan(0);
```

Reply:
I like the idea of adding in an expect expect to explicitly test this. However, I'd still like to leave the if block in. The reason being that the next line after the if block "const links = services.all(by.css('a.ng-star-inserted'));" will cause an Async Timeout error if no links are present (it seems to happen when you do a by css search on something that isn't there). Async Timeout errors will not only cause this test to fail but tend to also cause the next few tests to fail even if they should pass. To avoid that in the case of failure I think it would be fine to both have the explicit expect that tests that links are present and then also have the if to avoid the timeout that would occur afterwards if there are no links.

Reply:
It's also fine to just let this function end at this point as it will return a promise which is resolved in a function that expects if `links_tested` is greater than 0, which in this case will still be 0.